apps/recnet-api/src/modules/email/email.service.ts

```diff
@@ -0,0 +1,154 @@
+      const results = await Promise.all(promises);
+
+      const successCount = results.filter(
+        (result) => result.success && !result.skip
+      ).length;
+      const errorUserIds = results
+        .filter((result) => !result.success)
+        .map((result) => result.userId)
+        .filter((userId) => userId !== undefined) as string[];
+
+      // log the successful result to DB
+      await this.weeklyDigestCronLogRepository.updateWeeklyDigestCronLog(
+        cronLog.id,
+        { status: CronStatus.SUCCESS, result: { successCount, errorUserIds } }
+      );
+      logger.log(
+        `Finish weekly digest cron: ${successCount} emails sent, ${errorUserIds.length} errors`
+      );
+    } catch (error) {
+      logger.error(`Error in weekly digest cron: ${error}`);
+
+      // log the failed result to DB
+      const errorMsg = error instanceof Error ? error.message : "Unknown error";
+      await this.weeklyDigestCronLogRepository.updateWeeklyDigestCronLog(
+        cronLog.id,
+        { status: CronStatus.FAILURE, errorMsg }
+      );
+    }
+  }
+
+  public async sendTestEmail(userId: string): Promise<{ success: boolean }> {
+    const user = await this.userRepository.findUserById(userId);
+    const cutoff = getLatestCutOff();
+    return this.sendWeeklyDigest(user, cutoff);
+  }
+
+  private async sendWeeklyDigest(
+    user: DbUser,
+    cutoff: Date
+  ): Promise<SendMailResult> {
+    const followings = user.following.map(
+      (following: { followingId: string }) => following.followingId
+    );
+    const filter: RecFilterBy = {
+      userIds: followings,
+      cutoff,
+    };
+
+    // cap the number of recs to send in an email by MAX_REC_PER_MAIL
+    const dbRecs = await this.recRepository.findRecs(
+      1,
+      MAX_REC_PER_MAIL,
+      filter
+    );
+    const recs = dbRecs.map(this.getRecFromDbRec);
+    const recsGroupByTitle = groupBy(recs, (rec) => {
+      const titleLowercase = rec.article.title.toLowerCase();
+      const words = titleLowercase.split(" ").filter((w) => w.length > 0);
+      return words.join("");
+    });
+
+    // send email
+    const mailOptions = {
+      from: this.nodemailerConfig.user,
+      to: user.email,
+      subject: WeeklyDigestSubject(cutoff, this.appConfig.nodeEnv),
+      html: render(WeeklyDigest({ recsGroupByTitle })),
+    };
+
+    let result;
+    try {
+      result = await this.transporter.sendMail(user, mailOptions);
+    } catch (e) {
+      return { success: false, userId: user.id };
+    }
+
+    return result;
+  }
+
+  private getRecFromDbRec(dbRec: DbRec): Rec {
+    return {
+      ...dbRec,
+      cutoff: dbRec.cutoff.toISOString(),
+      user: {
+        id: dbRec.user.id,
+        handle: dbRec.user.handle,
+        displayName: dbRec.user.displayName,
+        photoUrl: dbRec.user.photoUrl,
+        affiliation: dbRec.user.affiliation,
+        bio: dbRec.user.bio,
+        numFollowers: dbRec.user.followedBy.length,
+      },
+    };
+  }
```

Comment:
Do you think we should expose this function from userRepository instead of re-defining here again?

Reply:
I think we can keep it here now.